### DISCUSSION TITLE
Homebrew install packages from disc

### DISCUSSION THREAD

**Discussion by mephmanx** [Homebrew/discussions#2423] [Linux]
All,

I am building a system install image and was wondering if there might be a way to save/cache homebrew and a list of packages to be installed from a disc/dvd/bluray/iso to save download time and bandwidth.  If I could build an installer with the core and a list of packages to build in to a repo or something...would really help.  Thank you in advance for any ideas or solutions.

**Comment by gromgit** [comment]
Easiest alternative I can think of:
1. Set up a live Homebrew instance exactly as you want it.
2. Archive `/home/linuxbrew/.linuxbrew` as a compressed tarball.

Then your installer simply unpacks the tarball into `/home/linuxbrew/.linuxbrew`, then runs:
```
eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"
brew update
brew upgrade
```

**Reply by mephmanx** [reply]
That is nice....Ill give it a shot.  Would I still be able to use the homebrew tools to add and remove packages as usual?  An approach like this wolnt break anything else?

**Reply by gromgit** [reply]
Yes. This works because Homebrew itself takes great pains to not touch anything outside its own hierarchy, so a tarball of `/home/linuxbrew/.linuxbrew` is a complete Homebrew snapshot.

**Comment by mephmanx** [comment]
So if I just untar that on a new VM it will see it as having homebrew installed and I could ad packages or whatever?

From: Adrian Ho ***@***.***>
Date: Monday, November 8, 2021 at 10:52 PM
To: Homebrew/discussions ***@***.***>
Cc: Chris Lyons ***@***.***>, Author ***@***.***>
Subject: Re: [Homebrew/discussions] Homebrew install packages from disc (Discussion #2423)

Yes. This works because Homebrew itself takes great pains to not touch anything outside its own hierarchy, so a tarball of /home/linuxbrew/.linuxbrew is a complete Homebrew snapshot.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Homebrew/discussions/discussions/2423#discussioncomment-1609956>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AMNAQ2UPWGWLYXDFLA3PW5TULCLHTANCNFSM5HT6NR7A>.




This email has been scanned by Inbound Shield™.

**Reply by gromgit** [reply]
Yes, though as I mentioned previously, you should `eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"` to set your environment up correctly, especially `PATH`.

Also consider running `brew doctor` after your first `brew update`, and fix any issues found, _before_ your first `brew upgrade`.

**Reply by mephmanx** [reply]
I have this coded up in my infra build and will let you know if everything works as expected but it looks good.  Very cleaver idea!!  Thank you very much!

**Comment by mephmanx** [comment]
I added this to my server build script and the only thing that was missing was something like this:

echo 'PATH=$PATH:/home/linuxbrew/.linuxbrew/bin' >> /opt/stack/.bash_profile


From: Adrian Ho ***@***.***>
Date: Monday, November 8, 2021 at 10:57 PM
To: Homebrew/discussions ***@***.***>
Cc: Chris Lyons ***@***.***>, Author ***@***.***>
Subject: Re: [Homebrew/discussions] Homebrew install packages from disc (Discussion #2423)

Yes, though as I mentioned previously, you should eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" to set your environment up correctly, especially PATH.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Homebrew/discussions/discussions/2423#discussioncomment-1609968>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AMNAQ2SZV6AH7GMK3AB3QSDULCLZPANCNFSM5HT6NR7A>.




This email has been scanned by Inbound Shield™.

**Reply by carlocab** [reply]
It wasn't missing. That's what
```
eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"
```
which @gromgit emphasised you needed to run is for.

**Comment by mephmanx** [comment]
I needed to add it to PATH…. Here is the full bit of my script:

telegram_notify $TELEGRAM_API $TELEGRAM_CHAT_ID "Found Homebrew cache, using cache for install...."
runuser -l root -c  'mkdir -p /home/linuxbrew'
runuser -l root -c  "tar -xf /tmp/homebrew-$CF_BBL_INSTALL_TERRAFORM_VERSION.tar -C /home/linuxbrew"
runuser -l stack -c  "echo 'PATH=$PATH:/home/linuxbrew/.linuxbrew/bin' >> /opt/stack/.bash_profile"
runuser -l stack -c  'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" >> /opt/stack/.bash_profile'
runuser -l stack -c  'brew update'
runuser -l stack -c  'brew upgrade'
runuser -l stack -c  'cd /opt/stack; source .bash_profile;'


From: Carlo Cabrera ***@***.***>
Date: Wednesday, November 10, 2021 at 7:29 PM
To: Homebrew/discussions ***@***.***>
Cc: Chris Lyons ***@***.***>, Author ***@***.***>
Subject: Re: [Homebrew/discussions] Homebrew install packages from disc (Discussion #2423)

It wasn't missing. That's what

eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"

which @gromgit<https://github.com/gromgit> emphasised you needed to run is for.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Homebrew/discussions/discussions/2423#discussioncomment-1623054>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AMNAQ2VTBB3UIMUNCXAVYBTULME43ANCNFSM5HT6NR7A>.




This email has been scanned by Inbound Shield™.

**Reply by carlocab** [reply]
If you run `/home/linuxbrew/.linuxbrew/bin/brew shellenv` without `brew` in your `PATH` you'll see that it emits a command that adds `brew` to your `PATH`.

**Reply by gromgit** [reply]
> ```
> runuser -l stack -c  "echo 'PATH=$PATH:/home/linuxbrew/.linuxbrew/bin' >> /opt/stack/.bash_profile"
> runuser -l stack -c  'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" >> /opt/stack/.bash_profile'
> runuser -l stack -c  'brew update'
> runuser -l stack -c  'brew upgrade'
> runuser -l stack -c  'cd /opt/stack; source .bash_profile;'
> ```

All that should be simply:
```
runuser -l stack -c 'echo eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" >> /opt/stack/.bash_profile'
runuser -l stack -c 'brew update; brew upgrade'
```
The first line adds the command I mentioned to the user's bash login startup file, while the second does the hard work. Since `runuser -l stack` runs its command in a login environment, the user's `.bash_profile` should automatically be sourced, thus setting up the shell's environment exactly as Homebrew likes it.

**Comment by mephmanx** [comment]
Maybe its just my system then…. The stack user didn’t have the ‘brew’ exec in path…I had to add it to path manually.

From: Adrian Ho ***@***.***>
Date: Wednesday, November 10, 2021 at 8:25 PM
To: Homebrew/discussions ***@***.***>
Cc: Chris Lyons ***@***.***>, Author ***@***.***>
Subject: Re: [Homebrew/discussions] Homebrew install packages from disc (Discussion #2423)

runuser -l stack -c  "echo 'PATH=$PATH:/home/linuxbrew/.linuxbrew/bin' >> /opt/stack/.bash_profile"

runuser -l stack -c  'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" >> /opt/stack/.bash_profile'

runuser -l stack -c  'brew update'

runuser -l stack -c  'brew upgrade'

runuser -l stack -c  'cd /opt/stack; source .bash_profile;'

All that should be simply:

runuser -l stack -c 'echo eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" >> /opt/stack/.bash_profile'

runuser -l stack -c 'brew update; brew upgrade'

The first line adds the command I mentioned to the user's bash login startup file, while the second does the hard work. Since runuser -l stack runs its command in a login environment, the user's .bash_profile should automatically be sourced, thus setting up the shell's environment exactly as Homebrew likes it.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Homebrew/discussions/discussions/2423#discussioncomment-1623216>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AMNAQ2QTQAVDZ36QW2ZKUBTULMLQZANCNFSM5HT6NR7A>.




This email has been scanned by Inbound Shield™.

**Reply by gromgit** [reply]
That's because:
```
runuser -l stack -c  'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" >> /opt/stack/.bash_profile'
```
causes the shell launched by `runuser` to execute the `eval` and append its output to your `.bash_profile`, but unless something went wrong with the `eval`, it would've output _nothing_. What that line really does is therefore:
1. Start a shell (let's call it `X`)
2. `eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" >> /opt/stack/.bash_profile`, which sets up Homebrew's environment in `X` but adds nothing to `.bash_profile`.
3. `X` exits, throwing away the Homebrew env settings from [2]

i.e. you just did a boatload of stuff for no net change.

That said, there's a big bug in the first line I mentioned. It should actually be:
```
runuser -l stack -c 'echo eval \"\$\(/home/linuxbrew/.linuxbrew/bin/brew shellenv\)\" >> /opt/stack/.bash_profile'
```
to account for the `runuser` shell's own re-`eval` of that command line.

**Comment by mephmanx** [comment]
Hmmm….ok, Ill try that.  I really appreciate your help, this has significantly reduced the install time for my infra.  Thank you for your help and guidance!  It means a lot to be able to ask for help on things like this and have someone respond, I haven’t experienced that on a lot of opensource libs I am using.

From: Adrian Ho ***@***.***>
Date: Wednesday, November 10, 2021 at 9:52 PM
To: Homebrew/discussions ***@***.***>
Cc: Chris Lyons ***@***.***>, Author ***@***.***>
Subject: Re: [Homebrew/discussions] Homebrew install packages from disc (Discussion #2423)

That's because:

runuser -l stack -c  'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" >> /opt/stack/.bash_profile'

causes the shell launched by runuser to execute the eval and append its output to your .bash_profile, but unless something went wrong with the eval, it would've output nothing. What that line really does is therefore:

  1.  Start a shell (let's call it X)
  2.  eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" >> /opt/stack/.bash_profile, which sets up Homebrew's environment in X but adds nothing to .bash_profile.
  3.  X exits, throwing away the Homebrew env settings from [2]

i.e. you just did a boatload of stuff for no net change.

That said, there's a big bug in the first line I mentioned. It should actually be:

runuser -l stack -c 'echo eval \"\$\(/home/linuxbrew/.linuxbrew/bin/brew shellenv\)\" >> /opt/stack/.bash_profile'

to account for the runuser shell's own re-eval of that command line.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Homebrew/discussions/discussions/2423#discussioncomment-1623465>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AMNAQ2TWEGCMREL7KOHFQALULMVVBANCNFSM5HT6NR7A>.




This email has been scanned by Inbound Shield™.
